triedb/firewood/database.go

```diff
@@ -0,0 +1,604 @@
+	proposalTree *ProposalContext
+}
+
+// New creates a new Firewood database with the given disk database and configuration.
+// Any error during creation will cause the program to exit.
+func New(diskdb ethdb.Database, config *Config) *Database {
+	if config == nil {
+		config = Defaults
+	}
+
+	fwConfig, path, err := validateConfig(config)
+	if err != nil {
+		log.Crit("firewood: error validating config", "error", err)
+	}
+
+	fw, err := ffi.New(path, fwConfig)
+	if err != nil {
+		log.Crit("firewood: error creating firewood database", "error", err)
+	}
+
+	currentRoot, err := fw.Root()
+	if err != nil {
+		log.Crit("firewood: error getting current root", "error", err)
+	}
+
+	return &Database{
+		fwDisk:      fw,
+		ethdb:       diskdb,
+		proposalMap: make(map[common.Hash][]*ProposalContext),
+		proposalTree: &ProposalContext{
+			Root: common.Hash(currentRoot),
+		},
+	}
+}
+
+func validateConfig(trieConfig *Config) (*ffi.Config, string, error) {
+	// Check that the directory exists
+	info, err := os.Stat(trieConfig.FilePath)
+	if err != nil {
+		return nil, "", fmt.Errorf("error checking database path: %w", err)
+	} else if !info.IsDir() {
+		return nil, "", fmt.Errorf("database path is not a directory: %s", trieConfig.FilePath)
+	}
+
+	// Append the filename to the path
+	if trieConfig.FileName == "" {
+		return nil, "", errors.New("no filename provided")
+	}
+	path := filepath.Join(trieConfig.FilePath, trieConfig.FileName)
+
+	// Check if the file exists
+	info, err = os.Stat(path)
+	exists := false
+	if err == nil {
+		if info.IsDir() {
+			return nil, "", fmt.Errorf("database file path is a directory: %s", path)
+		}
+		// File exists
+		log.Info("Database file found", "path", path)
+		exists = true
+	}
+
+	// Create the Firewood config from the provided config.
+	config := &ffi.Config{
+		Create:            !exists,                               // Use any existing file
+		NodeCacheEntries:  uint(trieConfig.CleanCacheSize) / 256, // TODO: estimate 256 bytes per node
+		Revisions:         trieConfig.Revisions,
+		ReadCacheStrategy: trieConfig.ReadCacheStrategy,
+		MetricsPort:       trieConfig.MetricsPort,
+	}
+
+	return config, path, nil
+}
+
+// Scheme returns the scheme of the database.
+// This is only used in some API calls
+// and in StateDB to avoid iterating through deleted storage tries.
+// WARNING: If cherry-picking anything from upstream that uses this,
+// it must be overwritten to use something like:
+// `_, ok := db.(*Database); if !ok { return "" }`
+// to recognize the Firewood database.
+func (db *Database) Scheme() string {
+	return rawdb.HashScheme
+}
+
+// Initialized indicates whether the root provided is the genesis root of the database.
+func (db *Database) Initialized(root common.Hash) bool {
+	// If the database is open, then the file is necessarily initialized.
+	// TODO: This would be more informative to have a flag to check that we've state synced later.
+	return db.fwDisk != nil
```

Comment:
why don't we check `fw.Root() != common.Hash`?, this looks like we actually always return `Initialized=true` since fwDisk is never nil after `New` but if we haven't commited anything `fw.Root` still can be empty (and actually not initialized)